lowtime/lowtime_rust/src/lib.rs

```diff
@@ -0,0 +1,111 @@
+use pyo3::prelude::*;
+
+use ordered_float::OrderedFloat;
+use pathfinding::directed::edmonds_karp::{
+    SparseCapacity,
+    Edge,
+    edmonds_karp
+};
+
+// for profiling
+use std::time::{
+    Instant,
+    Duration,
+};
+use log::info;
+
+
+#[pyclass]
+struct PhillipsDessouky {
+    #[pyo3(get)]
+    node_ids: Vec<u32>,
+    source_node_id: u32,
+    sink_node_id: u32,
+    edges_raw: Vec<((u32, u32), f64)>,
+}
+
+#[pymethods]
+impl PhillipsDessouky {
+    #[new]
+    fn new(
+        node_ids: Vec<u32>,
+        source_node_id: u32,
+        sink_node_id: u32,
+        edges_raw: Vec<((u32, u32), f64)>,
+    ) -> PyResult<Self> {
+        // Note: Intentionally nothing done here to profile data transfer latency
+
+        // ohjun: do we actually need to return this? This is unncessary data transfer from Rust->Python
+        // Instead, why not have a type that stores a pointer to a PhillipsDessouky, and we can just
+        // send the outer type? This way, Python can use the object to call max_flow and things, but
+        // we can avoid returning all this data here?
+        Ok(PhillipsDessouky {
+            node_ids: node_ids,
+            source_node_id: source_node_id,
+            sink_node_id: sink_node_id,
+            edges_raw: edges_raw,
+        })
```

Comment:
```suggestion
        Ok(PhillipsDessouky {
            node_ids,
            source_node_id,
            sink_node_id,
            edges_raw
        })
```

Constructor shorthand. See the [Rust book](https://doc.rust-lang.org/beta/book/ch05-01-defining-structs.html#using-the-field-init-shorthand).

Reply:
I think these aren't returned to Python BTW. It's just constructing the object from the Rust side, and Python will get a pointer to the Rust-side object. Whenever it access attributes (and it probably cannot since you didn't define `#[getter]`s for these attributes), it will call copy out the Rust-side data and convert it into Python objects.